lxd/ubuntupro/client.go

```diff
@@ -0,0 +1,261 @@
+		return fmt.Errorf("Invalid guest auto-attach setting %q", g)
+	}
+
+	return nil
+}
+
+// Settings contains Ubuntu Pro settings relevant to LXD.
+type Settings struct {
+	GuestAttach GuestAttachSetting `json:"guest_attach"`
+}
+
+// ubuntuAdvantageDirectory is the base directory for Ubuntu Pro related configuration.
+const ubuntuAdvantageDirectory = "/var/lib/ubuntu-advantage"
+
+// Client is our wrapper for Ubuntu Pro configuration and the Ubuntu Pro CLI.
+type Client struct {
+	guestAttachSetting GuestAttachSetting
+	monitor            fsmonitor.FSMonitor
+	pro                pro
+}
+
+// pro is an internal interface that is used for mocking calls to the pro CLI.
+type pro interface {
+	getProAPIGuestTokenDataAttributes(ctx context.Context) (*ProAPIGetGuestTokenV1DataAttributes, error)
+}
+
+// proCLI calls the actual Ubuntu Pro CLI to implement the interface.
+type proCLI struct{}
+
+// proAPIGetGuestTokenV1 represents the expected format of calls to `pro api u.pro.attach.guest.get_guest_token.v1`.
+// Not all fields are modelled as they are not required for guest attachment functionality.
+type proAPIGetGuestTokenV1 struct {
+	Result string `json:"result"`
+	Data   struct {
+		Attributes ProAPIGetGuestTokenV1DataAttributes `json:"attributes"`
+	} `json:"data"`
+	Errors []struct {
+		Title string
+	} `json:"errors"`
+}
+
+// ProAPIGetGuestTokenV1DataAttributes contains the expected fields of proAPIGetGuestTokenV1 that must be passed back to
+// the guest for pro attachment to succeed.
+type ProAPIGetGuestTokenV1DataAttributes struct {
+	Expires    string `json:"expires"`
+	GuestToken string `json:"guest_token"`
+	ID         string `json:"id"`
+}
+
+// getTokenJSON runs `pro api u.pro.attach.guest.get_guest_token.v1` and returns the result.
+func (proCLI) getProAPIGuestTokenDataAttributes(ctx context.Context) (*ProAPIGetGuestTokenV1DataAttributes, error) {
+	// Run pro guest attach command.
+	response, err := shared.RunCommandContext(ctx, "pro", "api", "u.pro.attach.guest.get_guest_token.v1")
+	if err != nil {
+		return nil, api.StatusErrorf(http.StatusServiceUnavailable, "Ubuntu Pro client command unsuccessful: %w", err)
+	}
+
+	var getGuestTokenResponse proAPIGetGuestTokenV1
+	err = json.Unmarshal([]byte(response), &getGuestTokenResponse)
+	if err != nil {
+		return nil, api.StatusErrorf(http.StatusInternalServerError, "Received unexpected response from Ubuntu Pro contracts server: %w", err)
+	}
+
+	if getGuestTokenResponse.Result != "success" {
+		if len(getGuestTokenResponse.Errors) > 0 && getGuestTokenResponse.Errors[0].Title != "" {
+			return nil, api.StatusErrorf(http.StatusServiceUnavailable, "Server returned %q when getting a guest token with error %q", getGuestTokenResponse.Result, getGuestTokenResponse.Errors[0].Title)
+		}
+
+		return nil, api.StatusErrorf(http.StatusServiceUnavailable, "Server returned %q when getting a guest token", getGuestTokenResponse.Result)
```

Comment:
We should stick with saying either "Ubuntu Pro contracts server" like above or just "server" here, but not mix them.